battlecode-manager/player_abstract.py

```diff
@@ -0,0 +1,99 @@
+from pathlib import Path
+import os
+import string
+import random
+import io
+import zipfile
+from shutil import copytree
+import sys
+
+
+# TODO Any reason why shutil.rmtree is not used for this?
+def delete_folder(path):
+    try:
+        for sub in path.iterdir():
+            if sub.is_dir():
+                delete_folder(sub)
+            else:
+                sub.unlink()
+        path.rmdir()
+    except Exception as e:
+        pass
+
+
+def random_key(length):
+    return ''.join([random.choice(string.ascii_letters + string.digits) for _ in range(length)])
+
+
+def extract_s3_bucket(bucket, key, destination_directory):
+    obj = bucket.Object(key)
+    with io.BytesIO(obj.get()["Body"].read()) as tf:
+        tf.seek(0)
+        with zipfile.ZipFile(tf, mode='r') as zipf:
+            zipf.extractall(path=destination_directory)
+
+
+def dos2unix(directory):
+    ''' Converts all .py and .sh files in the given directory (recursively) to use unix line endings '''
+    pathlist = list(Path(directory).glob("**/*.py"))
+    pathlist += list(Path(directory).glob("**/*.sh"))
+
+    for path in pathlist:
+        with open(str(path), 'r') as f:
+            x = f.read()
+        with open(str(path), 'w') as f:
+            f.write(x.replace('\r\n', '\n'))
+
+
+class AbstractPlayer:
+    def __init__(self, socket_file, working_dir, local_dir, s3_bucket, s3_key,
+                 player_key, player_mem_limit, player_cpu):
+        self.player_mem_limit = str(player_mem_limit) + 'mb'
+        self.player_key = player_key
+        self.socket_file = socket_file
+
+        if working_dir[-1] != "/":
+            working_dir += "/"
+
+        self.working_dir = Path(working_dir + random_key(20) + "/")
```

Comment:
This won't work on Windows, it needs to be:

`self.working_dir = os.path.abspath(os.path.join(working_dir, random_key(20)))`

You'll also need to remove the line above and change all the `str(self.working_dir.absolute())` to just `self.working_dir()`

Reply:
Huh. But in the sandbox.py:Sandbox class there is this code
~~~~
self.working_dir = Path(working_dir + random_key(20) + "/")
~~~~
which is code that runs outside of Docker if I understand things correctly as it runs before the player containers have been started. Or does that actually run inside of Docker?

Reply:
Hm... Though I suppose that might be running inside the manager container or something?

Reply:
Nah, the manager runs in docker as well, there's two layers :)

Reply:
Turtles all the way down!